File: modules/uml/app/model/matcher/UmlImplementationMatcher.java
package model.matcher;

import model.matching.Matcher;
import model.uml.UmlImplementation;

public class UmlImplementationMatcher extends Matcher<UmlImplementation, UmlImplementationMatch> {

  public UmlImplementationMatcher() {
    super(UmlImplementation::equals);
  }

  @Override
  protected UmlImplementationMatch instantiateMatch(UmlImplementation arg1, UmlImplementation arg2) {
    return new UmlImplementationMatch(arg1, arg2);
  }

}

File: app/model/tools/uml/matcher/UmlImplementationMatcher.scala
package model.tools.uml.matcher

import model.matching._
import model.tools.uml.UmlImplementation

final case class UmlImplementationMatch(
  matchType: MatchType,
  userArg: UmlImplementation,
  sampleArg: UmlImplementation
) extends Match[UmlImplementation]

object UmlImplementationMatcher extends Matcher[UmlImplementation, UmlImplementationMatch] {

  override protected def canMatch(i1: UmlImplementation, i2: UmlImplementation): Boolean =
    (i1.subClass == i2.subClass && i1.superClass == i2.superClass) || (i1.subClass == i2.superClass && i1.superClass == i2.subClass)

  override protected def instantiateMatch(ua: UmlImplementation, sa: UmlImplementation): UmlImplementationMatch = {

    val matchType = if (ua.subClass == sa.subClass && ua.superClass == sa.superClass) { MatchType.SUCCESSFUL_MATCH }
    else { MatchType.PARTIAL_MATCH }

    UmlImplementationMatch(matchType, ua, sa)
  }
}
